capellambse_context_diagrams/collectors/custom.py

```diff
@@ -0,0 +1,363 @@
+
+import capellambse.model as m
+
+from .. import _elkjs, context
+from . import generic, makers
+
+
+def _is_edge(obj: m.ModelElement) -> bool:
+    if hasattr(obj, "source") and hasattr(obj, "target"):
+        return True
+    return False
+
+
+def _is_port(obj: m.ModelElement) -> bool:
+    if obj.xtype.endswith("Port"):
+        return True
+    return False
+
+
+class CustomCollector:
+    """Collect the context for a custom diagram."""
+
+    def __init__(
+        self,
+        diagram: context.ContextDiagram,
+        params: dict[str, t.Any],
+    ) -> None:
+        self.diagram = diagram
+        self.target: m.ModelElement = self.diagram.target
+        if _is_port(self.target):
+            self.boxable_target = self.target.owner
+        elif _is_edge(self.target):
+            self.boxable_target = self.target.source.owner
+        else:
+            self.boxable_target = self.target
+        self.data = makers.make_diagram(diagram)
+        self.params = params
+        self.instructions = self.diagram._collect
+        self.repeat_instructions: dict[str, t.Any] = {}
+        self.repeat_depth: int = 0
+        self.visited: set[str] = set()
+        self.boxes: dict[str, _elkjs.ELKInputChild] = {}
+        self.edges: dict[str, _elkjs.ELKInputEdge] = {}
+        self.ports: dict[str, _elkjs.ELKInputPort] = {}
+        self.boxes_to_delete: set[str] = set()
+        if self.diagram._display_parent_relation:
+            self.edge_owners: dict[str, str] = {}
+            self.diagram_target_owners = list(
+                generic.get_all_owners(self.boxable_target)
+            )
+            self.common_owners: set[str] = set()
+        if self.diagram._unify_edge_direction != "NONE":
+            self.directions: dict[str, bool] = {}
+        self.min_heights: dict[str, dict[str, float]] = {}
+
+    def __call__(self) -> _elkjs.ELKInputData:
+        if _is_port(self.target):
+            self._make_port_and_owner(self.target)
+        else:
+            self._make_target(self.target)
+        if target_edge := self.edges.get(self.target.uuid):
+            target_edge.layoutOptions = copy.deepcopy(
+                _elkjs.EDGE_STRAIGHTENING_LAYOUT_OPTIONS
+            )
+        if not self.instructions:
+            return self._get_data()
+        if self.diagram._unify_edge_direction == "UNIFORM":
+            self.directions[self.boxable_target.uuid] = False
+        self._perform_instructions(self.target, self.instructions)
+        if self.diagram._display_parent_relation:
+            current = self.boxable_target
+            while (
+                current
+                and self.common_owners
+                and getattr(current, "owner", None) is not None
+                and not isinstance(current.owner, generic.PackageTypes)
+            ):
+                current = self._make_owner_box(
+                    current,
+                )
+                self.common_owners.discard(current.uuid)
```

Comment:
If `self.diagram._display_parent_relation` is set, every created box needs a while loop (in order to create their owners). We keep track of the common owners so that we only create as many owners for the target element as needed (we don't want to create all owners until root, just up until the highest common owner).

Once we implement derived components for `Custom Diagram` we could completely get rid of the default collector and just invoke the custom collector as we already do with context diagrams of `PhysicalPorts`. After that there won't be any more duplicate code.